Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1109 @@
+      let response = try state.framer.next(compressor: state.compressor)
+      self.state = .clientClosedServerOpen(state)
+      return response
+    case .clientOpenServerClosed, .clientClosedServerClosed:
+      throw self.assertionFailureAndCreateRPCErrorOnInternalError("Can't send response if server is closed.")
+    }
+  }
+  
+  private mutating func serverNextInboundMessage() -> [UInt8]? {
+    switch self.state {
+    case .clientOpenServerIdle(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientOpenServerIdle(state)
+      return request
+    case .clientOpenServerOpen(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientOpenServerOpen(state)
+      return request
+    case .clientOpenServerClosed(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientOpenServerClosed(state)
+      return request
+    case .clientClosedServerOpen(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientClosedServerOpen(state)
+      return request
+    case .clientClosedServerClosed(var state):
+      let request = state.inboundMessageBuffer.pop()
+      self.state = .clientClosedServerClosed(state)
+      return request
+    case .clientClosedServerIdle, .clientIdleServerIdle:
+      return nil
+    }
+  }
+}
+
+extension MethodDescriptor {
+  init?(fullyQualifiedMethod: String) {
+    let split = fullyQualifiedMethod.split(separator: "/")
+    guard split.count == 2 else {
+      return nil
+    }
+    self.init(service: String(split[0]), method: String(split[1]))
+  }
+}
+
+private enum GRPCHTTP2Keys: String {
+  case path = ":path"
+  case contentType = "content-type"
+  case encoding = "grpc-encoding"
+  case acceptEncoding = "grpc-accept-encoding"
+  case scheme = ":scheme"
+  case method = ":method"
+  case te = "te"
+  case status = ":status"
+  case grpcStatus = "grpc-status"
+  case grpcStatusMessage = "grpc-status-message"
+}
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension Metadata {
+  var path: MethodDescriptor? {
+    get {
+      self.firstString(forKey: .path)
+        .flatMap { MethodDescriptor(fullyQualifiedMethod: $0) }
+    }
+  }
+  
+  var contentType: ContentType? {
+    get {
+      self.firstString(forKey: .contentType)
+        .flatMap { ContentType(value: $0) }
+    }
+  }
+  
+  var encoding: CompressionAlgorithm? {
+    get {
+      self.firstString(forKey: .encoding).flatMap { CompressionAlgorithm(rawValue: $0) }
+    }
+  }
+  
+  var acceptedEncodings: [CompressionAlgorithm]? {
+    get {
+      self.firstString(forKey: .acceptEncoding)?
+        .split(separator: ",")
+        .compactMap { CompressionAlgorithm(rawValue: String($0)) }
+    }
+  }
+  
+  private func firstString(forKey key: GRPCHTTP2Keys) -> String? {
+    self[stringValues: key.rawValue].first(where: { _ in true })
+  }
+}
+
+extension HPACKHeaders {
```

Comment:
I'm slightly concerned about using all of these setters when constructing `HPACKHeaders` because they use `replaceOrAdd` which is O(n). Because we're creating the `HPACKHeaders` we know that there's no value to replace so a simple add would be absolutely fine.